File: LeetCode/5회차/B16-2-Add Two Numbers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        def cal(l1, l2, flag):
            if not l1:
                l1, l2 = l2, l1
            if not l1:
                return ListNode(1) if flag else None

            if l2:
                l1.val += l2.val
            l1.val += flag
            if l1.val >= 10:
                l1.val %= 10
                flag = 1
            else:
                flag = 0

            if l2:
                l1.next = cal(l1.next, l2.next, flag)
            else:
                l1.next = cal(l1.next, None, flag)
            return l1

        return cal(l1, l2, 0)


def make_list(li):
    root = node = ListNode()
    for num in li:
        node.next = ListNode(num)
        node = node.next
    return root.next


print(Solution().addTwoNumbers(make_list([2,4,9]), make_list([5,6,4,9])))

File: LeetCode/JavaScript/B16-2-Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  const root = new ListNode();
  let node = root;
  let carry = 0;

  while (l1 || l2 || carry) {
    let sum = 0;
    if (!!l1) {
      sum += l1.val;
      l1 = l1.next;
    }
    if (!!l2) {
      sum += l2.val;
      l2 = l2.next;
    }
    sum += carry;
    [sum, carry] = [sum % 10, Number(sum >= 10)];

    node.next = new ListNode(sum);
    node = node.next;
  }

  return root.next;
};
